### ISSUE TITLE
Set up of auto-correlation for http.createServer() does not cover all cases

### ISSUE THREAD

**Issue by 0x7f** [Microsoft/ApplicationInsights-node.js#311]
I was investigating an issue where the correlation context got lost in an [hapi](https://hapijs.com/) server. After digging through the code, I first expected their [domain-wrapper](https://github.com/hapijs/hapi/blob/master/lib/protect.js), but actually this works as expected.

The real issue was the way hapi calls `http.createServer`. The auto-collection code [expects](https://github.com/Microsoft/ApplicationInsights-node.js/blob/v0.17.2/AutoCollection/ServerRequests.ts#L45-L70) it being called like this (at least in latest release v0.17.2):
```
const handler = (req, res) => {...}
const server = http.createServer(handler);
```
but hapi does something like [this](https://github.com/hapijs/hapi/blob/v16.5.2/lib/connection.js#L89-L90):
```
const handler = (req, res) => {...}
const server = http.createServer();
server.on("request", handler);
```
which was introduced by [this patch](https://github.com/hapijs/hapi/commit/3b75e4f269fb4821a0d8abdc16180e503f73af6e) 3 years ago.

The most straight forward way to fix this would be to also patch the `server.on` function and hook into every event handler registering for the `request` (and `checkContinue`) event. What do you think?

**Comment by OsvaldoRosado** [comment]
Great catch! Thanks for spotting this!

**Comment by OsvaldoRosado** [comment]
Resolved with #312
